Add support for Broadlink BG1 devices

b4dpxl opened the pull request:
## Proposed change
This adds support for the British General Broadlink BG1 switches. Replaces PR #33399 


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
switch:
- platform: custom_broadlink
  host: 192.168.0.123
  mac: "24:df:a7:aa:bb:cc"
  type: bg1
  friendly_name: "Hall socket"
  timeout: 2

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: home-assistant/home-assistant.io#12540

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

b4dpxl committed:
Added support for BG1 devices

b4dpxl commented:
@felipediel this replaces https://github.com/home-assistant/core/pull/33399, after I managed to completely bork my dev branch :|

Goobiegoobers commented:
The code above works great for a single socket but have you tried with two sockets? Installing a second socket on the BG app shows as a sub device and doesn’t appear to have an IP address. It connected instantly to Alexa so there must be a way but I’m not clever enough to figure it out.

b4dpxl commented:
> The code above works great for a single socket but have you tried with two sockets? Installing a second socket on the BG app shows as a sub device and doesn’t appear to have an IP address. It connected instantly to Alexa so there must be a way but I’m not clever enough to figure it out.

@Goobiegoobers No, as I only have a single device. Does it not show up in your router's DHCP logs?

Goobiegoobers commented:
No it’s not there, and on the app it says IP unavailable, looks like it talks through the “main” device somehow

Goobiegoobers commented:
After further investigation I’ve found a second device definitely does not have its on IP address, it connects through what ever device is set up first. Quite a nice idea from BG to save on router devices I think.

b4dpxl commented:
> After further investigation I’ve found a second device definitely does not have its on IP address, it connects through what ever device is set up first. Quite a nice idea from BG to save on router devices I think.
> 
> I tried syncing one and then disconnecting it to sync the other but even with the “master” unplugged it still creates the 2nd as a “sub device”.

Unless these are doing something really odd, it must be connecting to your wireless network? There's nothing in the `broadlink` library I can see about connecting without it. Have you tried doing a discovery with the library to see what happens?

```
import broadlink
devices = broadlink.discover(timeout=5)
print(devices)
```

felipediel commented:
It looks like they're creating a mesh network. If confirmed, things will start to get complicated.

b4dpxl commented:
> It looks like they're creating a mesh network. If confirmed, things will start to get complicated.

If so, you should be able to see the network with a wifi scanner.

Goobiegoobers commented:
There must be a way to find it, as soon as it connected in the BG app I had a notification saying device found. 
@b4dpxl - apologies for my ignorance, where do I put the import broadlink code?

felipediel commented:
> > It looks like they're creating a mesh network. If confirmed, things will start to get complicated.
> 
> If so, you should be able to see the network with a wifi scanner.

Yes. Sorry, I don't have the devices to test.

felipediel commented:
@Goobiegoobers Type `python3` in the command line.

b4dpxl commented:
> > > It looks like they're creating a mesh network. If confirmed, things will start to get complicated.
> > 
> > 
> > If so, you should be able to see the network with a wifi scanner.
> 
> Yes. Sorry, I don't have the devices to test.

I only have one, and I don't see any evidence of a hidden network, though I've not done a thorough scan yet; I'll dig out one of my decent wireless cards at some point and have a scan.

Goobiegoobers commented:
I’m running HASS.OS so don’t have a command line. Is there another way to do it? Like I said I have two here plugged in and am happy to be a guinea pig to find the solution.

Goobiegoobers commented:
Ughh mine don’t work at all now haha. I’ve reset everything to as it was and I can’t even get one working on it’s own now 😞. Is it because I updated the firmware to v57053?

Snakesms commented:
> The code above works great for a single socket but have you tried with two sockets? Installing a second socket on the BG app shows as a sub device and doesn’t appear to have an IP address. It connected instantly to Alexa so there must be a way but I’m not clever enough to figure it out.

It might be that you can add 2nd socket as a sub devise for the 1st one, but its definitely not the default behaviour.

When I've added 2nd socket it's appeared as a 1st one exactly the same as a primary one in the app.
IP address is not visible in BG app for any devise in my case (1st or 2nd), but both easily searchable by MAC in the router.

Regarding this particular PR, I'm going to test it once figured out how to :)

Goobiegoobers commented:
> The code above works great for a single socket but have you tried with two sockets? Installing a second socket on the BG app shows as a sub device and doesn’t appear to have an IP address. It connected instantly to Alexa so there must be a way but I’m not clever enough to figure it out.
> 
> It might be that you can add 2nd socket as a sub devise for the 1st one, but its definitely not the default behaviour.
> When I've added 2nd socket it's appeared as a 1st one exactly the same as a primary one in the app.
> IP address is not visible in BG app for any devise in my case (1st or 2nd), but both easily searchable by MAC in the router.
> Regarding this particular PR, I'm going to test it once figured out how to :)

Snakesms, what firmware do you have running on your sockets? im on v57053 on ios and it does not give me a choice when adding a 2nd socket, it automatically sets it as a sub device.

Snakesms commented:
> Ughh mine don’t work at all now haha. I’ve reset everything to as it was and I can’t even get one working on it’s own now 😞. Is it because I updated the firmware to v57053?
> They are connected and working on everything except HA

I'm thinking there is something "wrong" in the latest Firmware, as one of my sockets used to work with the old firmware and become unavailable with the new one. I've got another socket with old firmware and avoiding firmware upgrade. Going to test if it works with old firmware or broken too

Goobiegoobers commented:
Keep me posted!

Snakesms commented:
> > > > It looks like they're creating a mesh network. If confirmed, things will start to get complicated.
> > > 
> > > 
> > > If so, you should be able to see the network with a wifi scanner.
> > 
> > 
> > Yes. Sorry, I don't have the devices to test.
> 
> I only have one, and I don't see any evidence of a hidden network, though I've not done a thorough scan yet; I'll dig out one of my decent wireless cards at some point and have a scan.

I did try to update 2 files preset in this PR in my running 101.5 version of HomeAssistant.
Result in the log is constantly:
"Platform broadlink not ready yet. Retrying in 60 seconds"

Respective config is switch section is:
  - platform: broadlink
    host: 192.168.X.X
    mac: "c8:f7:42:XX:XX:XX"
    type: bg1
    friendly_name: "Test"
    timeout: 2

Not sure if I'm missing anything and if it's possible to get some debugging info in the stable version of HomeAssistant to get more data or not

b4dpxl commented:
> Not sure if I'm missing anything and if it's possible to get some debugging info in the stable version of HomeAssistant to get more data or not

You can change the debug level in the configuration.yaml:
```yaml
logger:
  logs:
    homeassistant.components.broadlink: debug
```

Goobiegoobers commented:
@snakesms that config was working for me (before last night), does it not work for your socket with the earlier firmware either?

felipediel commented:
I figured out what is going on. [Check this out](https://youtu.be/O8mHPZv2iU4).

Snakesms commented:
> > Not sure if I'm missing anything and if it's possible to get some debugging info in the stable version of HomeAssistant to get more data or not
> 
> You can change the debug level in the configuration.yaml:
> 
> ```yaml
> logger:
>   logs:
>     homeassistant.components.broadlink: debug
> ```

Thanks, I'll try to update and catch something useful

Snakesms commented:
> @Snakesms that config was working for me (before last night), does it not work for your socket with the earlier firmware either?

Nope it doesn't.
"Test" socket is the one with old firmware.

Snakesms commented:
> I figured out what is going on. [Check this out](https://youtu.be/O8mHPZv2iU4).

That looks cool, but not what I'm getting in my case. In my case both sockets are primary and connected to WiFi with different IPs (but both doesn't work unfortunately

Goobiegoobers commented:
It’s exactly what I’m seeing in the BG app. But neither sockets are working either 😂

Snakesms commented:
> > Not sure if I'm missing anything and if it's possible to get some debugging info in the stable version of HomeAssistant to get more data or not
> 
> You can change the debug level in the configuration.yaml:
> 
> ```yaml
> logger:
>   logs:
>     homeassistant.components.broadlink: debug
> ```
I've enabled debug level but it doesn't add any new info at all.
The only output I can see in the log with regards to Broadlink component is:

> 2020-06-08 08:37:20 INFO (MainThread) [homeassistant.components.switch] Setting up > > switch.broadlink
2020-06-08 08:37:20 WARNING (MainThread) [homeassistant.components.switch] Platform broadlink not ready yet. Retrying in 180 seconds.
2020-06-08 08:37:20 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event system_log_event[L]: name=homeassistant.components.switch, message=['Platform broadlink not ready yet. Retrying in 180 seconds.'], level=WARNING, source=['helpers/entity_platform.py', 202], timestamp=1591601840.128314, exception=, count=1, first_occurred=1591601840.128314>

WIth components themselves I can see different behaviour between modified braodlink component & custom_broadlink (your original standalone component):
1) Modified always shows that its not ready
2) custom_broadlink shows:
"Log Details (ERROR)
Logger: custom_components.custom_broadlink.switch
Source: custom_components/custom_broadlink/switch.py:475
Integration: custom_broadlink
First occurred: 8:47:37 AM (3 occurrences)
Last logged: 8:56:24 AM
Auth failed: timed out"
&
"Logger: custom_components.custom_broadlink.switch
Source: custom_components/custom_broadlink/switch.py:153
Integration: custom_broadlink
First occurred: 8:47:01 AM (1 occurrences)
Last logged: 8:47:01 AM

Failed to connect to device"

b4dpxl commented:
> I figured out what is going on. [Check this out](https://youtu.be/O8mHPZv2iU4).

Something, something, MAGIC! 😄 Also, they don't seem sure if they developed it 😛 

![image](https://user-images.githubusercontent.com/26406101/84007932-ff1cec80-a968-11ea-8733-8ff3753e5792.png)

At some point I will pick up another socket, but I have no plans to do so in the near future. Also, the one I've got didn't appear in my app because it didn't like my network setup (dedicated, segregated VLAN for IoT devices).  But I suspect this is something for the upstream library to handle, not us.

b4dpxl commented:
> > > Not sure if I'm missing anything and if it's possible to get some debugging info in the stable version of HomeAssistant to get more data or not
> > 
> > 
> > You can change the debug level in the configuration.yaml:
> > ```yaml
> > logger:
> >   logs:
> >     homeassistant.components.broadlink: debug
> > ```
> 
> I've enabled debug level but it doesn't add any new info at all.
> The only output I can see in the log with regards to Broadlink component is:
> 
> > 2020-06-08 08:37:20 INFO (MainThread) [homeassistant.components.switch] Setting up > > switch.broadlink
> > 2020-06-08 08:37:20 WARNING (MainThread) [homeassistant.components.switch] Platform broadlink not ready yet. Retrying in 180 seconds.
> > 2020-06-08 08:37:20 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event system_log_event[L]: name=homeassistant.components.switch, message=['Platform broadlink not ready yet. Retrying in 180 seconds.'], level=WARNING, source=['helpers/entity_platform.py', 202], timestamp=1591601840.128314, exception=, count=1, first_occurred=1591601840.128314>
> 
> WIth components themselves I can see different behaviour between modified braodlink component & custom_broadlink (your original standalone component):
> 
> 1. Modified always shows that its not ready
> 2. custom_broadlink shows:
>    "Log Details (ERROR)
>    Logger: custom_components.custom_broadlink.switch
>    Source: custom_components/custom_broadlink/switch.py:475
>    Integration: custom_broadlink
>    First occurred: 8:47:37 AM (3 occurrences)
>    Last logged: 8:56:24 AM
>    Auth failed: timed out"
>    &
>    "Logger: custom_components.custom_broadlink.switch
>    Source: custom_components/custom_broadlink/switch.py:153
>    Integration: custom_broadlink
>    First occurred: 8:47:01 AM (1 occurrences)
>    Last logged: 8:47:01 AM
> 
> Failed to connect to device"

There's not necessarily any debugging in the code to help unfortunately 🤷

Snakesms commented:
> There's not necessarily any debugging in the code to help unfortunately 🤷
Could you advise please:
What firmware version do you have that does work with your custom component?
And what app did you use to connect it to the Broadlink cloud?
Android/IOS & BG Home/BroadLink

Goobiegoobers commented:
I had it working prior to 57053

b4dpxl commented:
> > There's not necessarily any debugging in the code to help unfortunately 🤷
> > Could you advise please:
> > What firmware version do you have that does work with your custom component?
> > And what app did you use to connect it to the Broadlink cloud?
> > Android/IOS & BG Home/BroadLink

Firmware version, no idea. I initially linked it using the iOS app, but it never actually showed up, and I removed the app. In theory you can pair them using the `broadlink` library [directly too](https://github.com/mjg59/python-broadlink#example-use), but my network isn't flat.

felipediel commented:
> > I figured out what is going on. [Check this out](https://youtu.be/O8mHPZv2iU4).
> 
> Something, something, MAGIC! smile Also, they don't seem sure if they developed it stuck_out_tongue
> 
> ![image](https://user-images.githubusercontent.com/26406101/84007932-ff1cec80-a968-11ea-8733-8ff3753e5792.png)
> 
> At some point I will pick up another socket, but I have no plans to do so in the near future. Also, the one I've got didn't appear in my app because it didn't like my network setup (dedicated, segregated VLAN for IoT devices). But I suspect this is something for the upstream library to handle, not us.

Whaaat?! Hahahahaha :smile:

felipediel commented:
Okay, we have two problems here:
1. Our library does not fully support the FastCon technology.
2. The BG app does not support the new firmware for these devices.

None of them is up to Home Assistant to resolve. Your code looks good to me, although I don't have the device to test it.

I still need to finish developing the config flow before we merge your code. It will need minor adaptations, but they are simple and I will help you as soon as everything is ready.

felipediel commented:
As for support for FastCon technology and subdevices, I can help to decrypt and analyze the communication once the BG app is working properly. But this needs to be done in the [library](https://github.com/mjg59/python-broadlink) and not here. So someone please open an issue there and I will be happy to contribute with my tools.

Goobiegoobers commented:
Hi felipediel, 
So you know the BG app works as BG have intended with two sockets connected. There is a master and a sub, they connect to Alexa instantly and I can also control from Siri directly. I messaged BG asking if it’s possible to have two masters and the reply was yes (although I tried what was suggested and it didn’t work). I will try that again tomorrow. 
The only place is it not working is with HA. And where I had one socket working before I cannot get either to work.

Goobiegoobers commented:
I have good news.... I think.
I have managed to connect two sockets as masters with separate IP addresses! It is very much doable!
Unfortunately I’m still in the position where neither work, but a step forward!

Goobiegoobers commented:
> > > > > It looks like they're creating a mesh network. If confirmed, things will start to get complicated.
> > > > 
> > > > 
> > > > If so, you should be able to see the network with a wifi scanner.
> > > 
> > > 
> > > Yes. Sorry, I don't have the devices to test.
> > 
> > 
> > I only have one, and I don't see any evidence of a hidden network, though I've not done a thorough scan yet; I'll dig out one of my decent wireless cards at some point and have a scan.
> 
> I did try to update 2 files preset in this PR in my running 110.5 version of HomeAssistant.
> Result in the log is constantly:
> "Platform broadlink not ready yet. Retrying in 60 seconds"
> 
> Respective config is switch section is:
> 
> * platform: broadlink
>   host: 192.168.X.X
>   mac: "c8:f7:42:XX:XX:XX"
>   type: bg1
>   friendly_name: "Test"
>   timeout: 2
> 
> Not sure if I'm missing anything and if it's possible to get some debugging info in the stable version of HomeAssistant to get more data or not

@Snakesms I've just realised a bit that is wrong. your platform should be custom_broadlink

Give that a go and let us know how you get on.

Goobiegoobers commented:
> > After further investigation I’ve found a second device definitely does not have its on IP address, it connects through what ever device is set up first. Quite a nice idea from BG to save on router devices I think.
> > I tried syncing one and then disconnecting it to sync the other but even with the “master” unplugged it still creates the 2nd as a “sub device”.
> 
> Unless these are doing something really odd, it must be connecting to your wireless network? There's nothing in the `broadlink` library I can see about connecting without it. Have you tried doing a discovery with the library to see what happens?
> 
> ```
> import broadlink
> devices = broadlink.discover(timeout=5)
> print(devices)
> ```

I finally got python on my laptop (separate to HA)
I get:
> import broadlink
> devices = broadlink.discover(timeout=5)
> print(devices)
> [ ]
attached image shows both sockets on my network with dedicated ips

![image](https://user-images.githubusercontent.com/29177370/84131190-25f82300-aa3c-11ea-9b39-cda4238dcc92.png)

@b4dpxl , does your single socket work currently on HA or has yours lost connection too? I don't understand why i cannot connect a single socket anymore. they're separate entities now so it shouldn't be an issue (unless they've changed something in their firmware to affect us of course)

b4dpxl commented:
Mine’s working fine, though discover didn’t find it. You could try increasing the timeout from 5. That might only work if they’re in setup mode?

Snakesms commented:
> b4dpxl
When devices in setup mode they are not connected to WiFi yet so can't be descoverable too

Goobiegoobers commented:
> Mine’s working fine, though discover didn’t find it. You could try increasing the timeout from 5. That might only work if they’re in setup mode?

It's interesting that your one is working, i thought everyone's had failed. if you don't have the app i guess you haven't updated the socket firmware? that's the only thing i can think that could be stopping the rest of us. I've messaged BG to see if there is a way to revert the firmware.

felipediel commented:
@Goobiegoobers Let's do a test. Try this:
```python3
import broadlink as blk
d1 = blk.discover(timeout=10, discover_ip_address="192.168.0.44")
d2 = blk.discover(timeout=10, discover_ip_address="192.168.0.46")
print(d1, d2)
```

felipediel commented:
Test number 2:
```python3
import broadlink as blk
d1 = blk.gendevice(0x51E3, ("192.168.0.44", 80), b"\x117\x97B\xf7\xc8')
d2 = blk.gendevice(0x51E3, ("192.168.0.46", 80), b"\x97C\x97B\xf7\xc8")
print(d1.auth())
print(d2.auth())

b4dpxl commented:
By way of comparison (I was curious):
```python
>>> x = blk.discover(timeout=20, discover_ip_address="192.168.2.186")
>>> x
[<broadlink.bg1 object at 0x7f21d4741ed0>]
>>> x[0].auth()
True
>>> x[0].get_state()
{'pwr': 1, 'pwr1': 0, 'pwr2': 1, 'maxworktime': 0, 'maxworktime1': 0, 'maxworktime2': 0, 'idcbrightness': 100}
```

felipediel commented:
Could you guys provide me a link to your devices [here](https://www.bgelectrical.uk/smart-sockets)? I am working on an update that will display the model to the user.

b4dpxl commented:
Mine is the BG800 - https://www.bgelectrical.uk/wiring-devices/moulded/smart-sockets#822/HC-01. As far as I can tell, the only difference between the 800 and 900 is the shape of the socket's faceplate; the 800 range has a "slim" rounded profile, whereas the 900 range is square.

Goobiegoobers commented:
Test 1
![image](https://user-images.githubusercontent.com/29177370/84179721-4ea10c80-aa7e-11ea-9a0c-953ad922b974.png)
Test 2
![image](https://user-images.githubusercontent.com/29177370/84179862-827c3200-aa7e-11ea-8b10-cef5febd241f.png)

Goobiegoobers commented:
to confirm they're on the network
![image](https://user-images.githubusercontent.com/29177370/84179995-b5bec100-aa7e-11ea-830f-92929e82333b.png)

felipediel commented:
@Goobiegoobers Your devices are blocked for local authentication. I thought this feature was unique to the RM4 series, but it also appears to be in the latest firmware for BG devices. The only solution is to reset the device and try to add it to the network again, interrupting the configuration before cloud authentication is performed.

This might work:
1. Factory reset the device.
2. Add the device to the local network.
2. STOP. Do not finish the setup. Close the app.

Then repeat the discovery and authentication attempts and tell me if it worked.

Goobiegoobers commented:
@felipediel you genius!
![image](https://user-images.githubusercontent.com/29177370/84184978-38974a00-aa86-11ea-8449-cfadebdadadb.png)
working in HA too!!!

felipediel commented:
@Goobiegoobers Great! Thank you! Is yours the BG800 too?

Goobiegoobers commented:
> @Goobiegoobers Great! Thank you! Is yours the BG800 too?

I have the BG900 
https://www.bgelectrical.uk/wiring-devices/moulded/smart-sockets#922/HC-01

strongy-uk commented:
just used the same method as @Goobiegoobers and can confirm mines working aswell.

Cheers Guys!! now to buy more (Y)

Snakesms commented:
> @felipediel you genius!
> ![image](https://user-images.githubusercontent.com/29177370/84184978-38974a00-aa86-11ea-8449-cfadebdadadb.png)
> working in HA too!!!
Hi Goobiegoobers,

What HA component do you currently use?
I suppose current PR is not merged into the master yet

@felipediel 
Thanks a lot, I can confirm I've followed the same approach and local authentication does work for me as well

Snakesms commented:
And when everyone breathed a sigh of relief...
here is another challenge for you :)

How to add 2nd device to the WiFi network?

Once you followed the steps:

...
...
3. STOP. Do not finish the setup. **Close the app.**

And opened it again trying to add 2nd Socket, you'll see following picture
![SmartSelect_20200616-210549_BroadLink](https://user-images.githubusercontent.com/39484499/84824304-9d6b1b00-b017-11ea-8eaf-2d44accab548.jpg)

where "Socket" visible on the screenshot is actually 1st one.
If by any chance you press on the Socket and complete setup - your Socket1 will stop working and disappear from HA.
Now you'll be able to add Socket2 and close the app without adding it to the app.

And we are getting into the loop where only last Socket is not added to the app so Does work in HA

Goobiegoobers commented:
> @felipediel you genius!
> 
> working in HA too!!!
> Hi Goobiegoobers,
> 
> What HA component do you currently use?
> I suppose current PR is not merged into the master yet
> @felipediel
> Thanks a lot, I can confirm I've followed the same approach and local authentication does work for me as well

Sorry, What do you mean what component?

Goobiegoobers commented:
> And when everyone breathed a sigh of relief...
> here is another challenge for you :)
> How to add 2nd device to the WiFi network?
> Once you followed the steps:
> ...
> ...
> 3. STOP. Do not finish the setup. Close the app.
> And opened it again trying to add 2nd Socket, you'll see following picture
> 
> where "Socket" visible on the screenshot is actually 1st one.
> If by any chance you press on the Socket and complete setup - your Socket1 will stop working and disappear from HA.
> Now you'll be able to add Socket2 and close the app without adding it to the app.
> And we are getting into the loop where only last Socket is not added to the app so Does work in HA

It's fiddly but doable. 
Add both sockets in the app and assign local IP addresses for each of them on your router. then delete the sockets from the app and carry out the instructions @felipediel explained

b4dpxl commented:
Has anyone tried adding the devices without using the app, just using the python library? 

https://github.com/mjg59/python-broadlink/blob/master/README.md#example-use

Snakesms commented:
> Has anyone tried adding the devices without using the app, just using the python library?
> 
> https://github.com/mjg59/python-broadlink/blob/master/README.md#example-use

I didn't and even didn't know such option exist. Will try today

Snakesms commented:
> > And when everyone breathed a sigh of relief...
> > here is another challenge for you :)
> > How to add 2nd device to the WiFi network?
> > Once you followed the steps:
> > ...
> > ...
> > 3. STOP. Do not finish the setup. Close the app.
> > And opened it again trying to add 2nd Socket, you'll see following picture
> > where "Socket" visible on the screenshot is actually 1st one.
> > If by any chance you press on the Socket and complete setup - your Socket1 will stop working and disappear from HA.
> > Now you'll be able to add Socket2 and close the app without adding it to the app.
> > And we are getting into the loop where only last Socket is not added to the app so Does work in HA
> 
> It's fiddly but doable.
> Add both sockets in the app and assign local IP addresses for each of them on your router. then delete the sockets from the app and carry out the instructions @felipediel explained

It looks like you didn't get my point...

- Did you try to add 2 socket yourself and it was successful ?
- It have nothing to do with Local IP

Will try again, the point is:
If you didn't finish setup process for Socket1 in the app, when you'll try to add Socket2 in the app - you won't see it in the list, instead - you'll always see Socket1 in device list and App will push you to complete setup

Goobiegoobers commented:
Yes I have two working. 
I used the AP config to set them up instead of smart config, maybe try that? 
I was getting messages trying to get me to connect the other one all the way through but had to ignore them, that would connect them and also put them into the mesh mode.

Snakesms commented:
Thanks for answering. Yes, AP mode might be an option

Snakesms commented:
Hi @b4dpxl 

Have you seen that this PR didn't pass all required checks so far?
We are appreciate your business and understand you might be busy with other stuff, but it's 95% completed and is very sad that last bit take sooo long

Thanks and regards

b4dpxl commented:
> Hi @b4dpxl
> 
> Have you seen that this PR didn't pass all required checks so far?
> We are appreciate your business and understand you might be busy with other stuff, but it's 95% completed and is very sad that last bit take sooo long
> 
> Thanks and regards

Hi @Snakesms. I'm waiting on @felipediel to finish another PR which will require some changes to this PR first, which is why it's not being pushed.
> I still need to finish developing the config flow before we merge your code. It will need minor adaptations, but they are simple and I will help you as soon as everything is ready.

Also, both the failed checks just show `Bash exited with code '4'.` in the details. It passed all the local checks (as per the dev. documentation) so I have no idea how to fix that issue :man_shrugging:

g189ffin commented:
Hey guys, I managed to connect the second switch on to the network and the AP mode definitely help! I've got two master sockets now which is great. I also went into my configuration.yaml for home assistant and added them on there and restarted my HA. I can now see 2 pairs of sockets, but only one controllable and the other one greyed out (second socket in my configuration.yaml). I've got it set up as:

switch:
  - platform: custom_broadlink
    host: 192.168.1.168
    mac: "c8:f7:42:97:54:06"
    type: bg1
    friendly_name: "Kitchen Socket"
    timeout: 2

  - platform: custom_broadlink
    host: 192.168.1.169
    mac: "c8:f7:42:86:93:cb"
    type: bg1
    friendly_name: "Living Room Socket"
    timeout: 2

Can't see why the second socket is greyed out in HA. Any ideas?

Goobiegoobers commented:
Can you ping the second socket?

g189ffin commented:
Yea I am able to ping to it but I think I'm having that issue that you had @Goobiegoobers with authentication failing in python. How did you fix it? 

I used AP mode to connect it to the network otherwise the smart setup just takes over and creates the new device as a sub.

Goobiegoobers commented:
Message from Felipedel solved it for me (2 months ago)

https://github.com/home-assistant/core/pull/36449#issuecomment-641483409

g189ffin commented:
I'm in guys!! I used the broadlink.setup(ssid, password, mode) to set up my device

> > https://github.com/mjg59/python-broadlink/blob/master/README.md#example-use

By using that I was able to connect my device to the network. The issue I was facing was that if:
- I used the BG Home app in smart mode, it sets the second socket as a sub
- I used the BG Home app in AP mode, when I run `devices[0].auth()` I get `False`. This works with the app very well but not with Home Assistant

When I finally used the python library to set it up, it was the barebones I needed it to do i.e. connect it to the network and when i ran `devices[0].auth`, I got `True` and now its connected on Home Assistant but not in the app. It is also not controllable through google home unless you link Home assistant as your integration into Google Home which is fine for my use.

Snakesms commented:
Hi g189ffin,
Based on my experience, there is no way to add socket to both: Ha & App.
So you have to chose what you prefer.
The best way probably would be to add to HA and export it to Google Home later as HA integration and not directly from the BG app

Regards

mebrew17 commented:
Hi I am a HA noobie. I have followed eugene eschava's broadlink mqtt scheme but have only got to : " root@UbuntuDesktop:/home/mebrew/broadlink/broadlink-mqtt# python3 mqtt.py
[2020-08-25 16:24:18,023] DEBUG Connected to 'BG1' Broadlink device at '192.168.1.107' (MAC 13:4b:97:42:f7:c8) and started listening for commands at MQTT topic having prefix 'broadlink/' 
[2020-08-25 16:24:18,555] DEBUG Connected to MQTT broker, subscribing to topic broadlink/# "
but can get no further.  
Any tips on how to proceed and configure into Home Assisatnt.

thank you

felipediel commented:
@b4dpxl I think we are ready for this update. Please rebase it --onto dev branch so I can help you.

## How to rebase:
1. Go to your local branch, e.g:
`cd Projects/home-assistant`
`git checkout broadlinkbg1`

2. If you don't have an upstream yet:
`git remote add upstream https://github.com/home-assistant/core.git`

3. Fetch upstream:
`git fetch upstream`

4. Check your changes. How many commits?
`git log`

5. Rebase
`git rebase --onto upstream/dev HEAD~NUMBER_OF_COMMITS`

6. Open the folder in VS code and solve conflicts. Add BG1 to:
- [DOMAINS_AND_TYPES](https://github.com/home-assistant/core/blob/7bc273b182b8ff94172fe653206a0c7190ed85a6/homeassistant/components/broadlink/const.py#L11)
- [async_setup_entry()](https://github.com/home-assistant/core/blob/7bc273b182b8ff94172fe653206a0c7190ed85a6/homeassistant/components/broadlink/switch.py#L107)
- [get_update_manager()](https://github.com/home-assistant/core/blob/7bc273b182b8ff94172fe653206a0c7190ed85a6/homeassistant/components/broadlink/updater.py#L19)

7. Implement a BroadlinkUpdateManager to handle updates.

8. Test your changes locally.
`source venv/bin/activate`
`hass -c config`

9. Test if local tests pass:
`pytest tests/components/broadlink`

10. Commit your changes and force push to origin.
`git add .`
`git commit -m 'Add support for BG1 devices'`
`git push origin broadlinkbg1 --force`

Ping me when you are done.

b4dpxl commented:
> @b4dpxl I think we are ready for this update. Please rebase it --onto dev branch so I can help you.

Many thanks for the detailed instructions @felipediel - It's going to be a couple of weeks before I can look at this, but I'll get it in ASAP.

bdraco commented:
I'm going to mark this as a draft since its going to be a few weeks before the author can come back to this.  Please mark this ready for review once its ready for review.

dodavies commented:
Does this then mean it will appear in the beta within hass io when the author carries out the requested tasks and then eventually in the non beta ?

fryc88 commented:
Good. I'm really waiting for it and keeping my fingers crossed to get it working! That's really great job - i think it's best quality of the socket for their price.

b4dpxl commented:
> @b4dpxl I think we are ready for this update. Please rebase it --onto dev branch so I can help you.
> 
> ## How to rebase:
> 1. Go to your local branch, e.g:
>    `cd Projects/home-assistant`
>    `git checkout broadlinkbg1`
> 2. If you don't have an upstream yet:
>    `git remote add upstream https://github.com/home-assistant/core.git`
> 3. Fetch upstream:
>    `git fetch upstream`
> 4. Check your changes. How many commits?
>    `git log`
> 5. Rebase
>    `git rebase --onto upstream/dev HEAD~NUMBER_OF_COMMITS`
> 6. Open the folder in VS code and solve conflicts.
>    Use `git add .` and `git rebase --continue` to continue.
> 7. Add BG1 to:
> 
> * [DOMAINS_AND_TYPES](https://github.com/home-assistant/core/blob/7bc273b182b8ff94172fe653206a0c7190ed85a6/homeassistant/components/broadlink/const.py#L11)
> * [async_setup_entry()](https://github.com/home-assistant/core/blob/7bc273b182b8ff94172fe653206a0c7190ed85a6/homeassistant/components/broadlink/switch.py#L107)
> * [get_update_manager()](https://github.com/home-assistant/core/blob/7bc273b182b8ff94172fe653206a0c7190ed85a6/homeassistant/components/broadlink/updater.py#L19)
> 
> 1. Implement a BroadlinkUpdateManager to handle updates.
> 2. Test your changes locally.
>    `source venv/bin/activate`
>    `hass -c config`
> 3. Test if local tests pass:
>    `pytest tests/components/broadlink`
> 4. Commit your changes and force push to origin.
>    `git add .`
>    `git commit -m 'Add support for BG1 devices'`
>    `git push origin broadlinkbg1 --force`
> 
> Ping me when you are done.

Hi @felipediel. `git rebase ... HEAD~27280` gave an error about "invalid upstream", although `get rebase ... <last_commit_id_from_git_log>` worked. But what am I supposed to do here?

```
        added by them:   app/DeviceTracker.py
        deleted by us:   app/EventBus.py
        deleted by us:   app/actor/HueTrigger.py
        deleted by us:   app/observer/TomatoDeviceScanner.py
        deleted by us:   app/util.py
        deleted by us:   start.py
```

what do I `git add`, and what do I `git rm`? 

Thanks

felipediel commented:
You need to use HEAD~NUMBER_OF_COMMITS. Let's say your only commit is "Added support for BG1 devices". So you `git rebase --onto upstream/dev HEAD~1`. I suggest you to `git log` before to check the commits.

b4dpxl commented:
what am I actually looking for in git log? I've only ever used git for my personal projects, with just me working on it. 

After a `git fetch upstream` there are lots of commits in `git log`, some older and some newer than mine (which isn't in the log). How do I determine `NUMBER_OF_COMMITS`?

felipediel commented:
It may be easier for you to copy the bg1 class to an updated branch and start over.

dodavies commented:
how does this get resolved, I'm super keen to have this in beta and stable?

dodavies commented:
still keen to understand how this gets resolved as this is a key integration in the UK

b4dpxl commented:
> still keen to understand how this gets resolved as this is a key integration in the UK

I need time to rework it given the changes to the core module. However that’s not likely to be any time soon due to personal circumstances. If someone else wants to have a go it should be relatively straight forward.

mebrew17 commented:
Agreed. I would like to get a fleet of these sockets as soon as the integration is alive.

> On 12 Oct 2020, at 18:43, Dave Davies <notifications@github.com> wrote:
> 
> 
> still keen to understand how this gets resolved as this is a key integration in the UK
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/home-assistant/core/pull/36449#issuecomment-707258208>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AORPXWK2BDOZWNKC677RGCTSKM53ZANCNFSM4NS2CJ5Q>.
>

mebrew17 commented:
Thank you for the update. I appreciate that you do this whenever you got time. I wish I had the IT skills to help.

> On 12 Oct 2020, at 18:46, b4dpxl <notifications@github.com> wrote:
> 
> 
> still keen to understand how this gets resolved as this is a key integration in the UK
> 
> I need time to rework it given the changes to the core module. However that’s not likely to be any time soon due to personal circumstances. If someone else wants to have a go it should be relatively straight forward.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/home-assistant/core/pull/36449#issuecomment-707259544>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AORPXWN3BNPS67HDE3ZFPEDSKM6GNANCNFSM4NS2CJ5Q>.
>

dodavies commented:
If someone wants to give me steps to get this past the post I'm happy to do so as from what I can tell its a matter of git stuff not code stuff ?

b4dpxl commented:
> If someone wants to give me steps to get this past the post I'm happy to do so as from what I can tell its a matter of git stuff not code stuff ?

It needed code changes. I should be in a position to push a new PR over the weekend.

dodavies commented:
That is excellent news, with a colleague we tried but got as far as Implement a BroadlinkUpdateManager to handle updates.
But now I am lost with no code skill.
I look forward to your PR and thanks again for doing it !

JimZAH commented:
I also would love to see this integration make it. The sockets are fairly priced for the quality and it means I can remove bulky Sonoff plugs. Sadly I don't have the skill set in Python to assist.

b4dpxl commented:
https://github.com/home-assistant/core/pull/42314 created to replace this.
